packages/node/src/policy-gen/to-policy.js

```diff
@@ -132,63 +163,62 @@ export const buildModuleRecords = (
     throw new TypeError('Could not find entry compartment; this is a bug')
   }
 
-  const compartmentRenames = freeze(renames)
+  const compartmentRenames = freeze({ ...renames })
+
+  const entrypointPath = toPath(entrypoint)
 
-  /** @type {MissingModuleMap} */
-  const missingModules = new Map()
-  const contexts = entries(compartmentMap.compartments)
-    // TODO: warn about this? how frequently does this occur?
-    // likewise: can something be in sources but not in the compartment map?
-    .filter(([compartmentName]) => compartmentName in sources)
-    .map(
-      ([compartmentName, compartment]) =>
-        /** @type {[string, PolicyGeneratorContext]} */ ([
+  const contexts = entries(compartmentMap.compartments).reduce(
+    (acc, [compartmentName, compartment]) => {
+      if (compartmentName in sources) {
+        acc.push([
           compartmentName,
           PolicyGeneratorContext.create(
             compartment,
             compartmentRenames,
             lmrCache,
             {
-              isEntry: entryCompartment === compartment,
+              rootModule:
+                compartment === entryCompartment ? entrypointPath : undefined,
               readPowers,
               isBuiltin,
-              missingModules,
               log,
             }
           ),
         ])
-    )
-
-  let moduleRecords = contexts
-    .flatMap(([compartmentName, context]) => {
-      if (!(compartmentName in sources)) {
-        // "should never happen"™
-        throw new ReferenceError(
-          `Could not find corresponding source for ${compartmentName}; this is a bug`
-        )
       }
+      return acc
+    },
+    /**
+     * @type {[
+     *   compartmentName: string,
+     *   context: PolicyGeneratorContext<any>,
+     * ][]}
+     */ ([])
+  )
 
-      const compartmentSources = sources[compartmentName]
+  const moduleRecords = contexts.reduce((acc, [compartmentName, context]) => {
```

Comment:
(nit) The benefit of flatMap over reduce is with reduce you can do absolutely anything, with flatMap you know you're getting 1 or more items per each item in the original list. It narrows down what could go wrong on changes.
The difference is negligible, but this IMHO is not an improvement

Reply:
was thinking in terms of complexity, not understandability 😄

given this is marked `nit` I'm likely to leave it as-is

Reply:
The brainpower it takes to understand the codebase is such that this is too small to affect it. I'd argue against a reduce if it was in a crud app designed for juniors to maintain.